idx/memory/memory.go

```diff
@@ -0,0 +1,475 @@
+package memory
+
+import (
+	"flag"
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/raintank/met"
+	"github.com/raintank/metrictank/idx"
+	"github.com/raintank/worldping-api/pkg/log"
+	"github.com/rakyll/globalconf"
+	"gopkg.in/raintank/schema.v1"
+)
+
+var (
+	idxOk             met.Count
+	idxFail           met.Count
+	idxAddDuration    met.Timer
+	idxGetDuration    met.Timer
+	idxListDuration   met.Timer
+	idxFindDuration   met.Timer
+	idxDeleteDuration met.Timer
+
+	Enabled bool
+)
+
+func ConfigSetup() {
+	memoryIdx := flag.NewFlagSet("memory-idx", flag.ExitOnError)
+	memoryIdx.BoolVar(&Enabled, "enabled", false, "")
+	globalconf.Register("memory-idx", memoryIdx)
+}
+
+type Tree struct {
+	Items map[string]*Node
+}
+
+type Node struct {
+	Path     string
+	Children []string
+	Leaf     bool
+}
+
+func (n *Node) String() string {
+	t := "branch"
+	if n.Leaf {
+		t = "leaf"
+	}
+	return fmt.Sprintf("%s - %s", t, n.Path)
+}
+
+// Implements the the "MetricIndex" interface
+type MemoryIdx struct {
+	sync.Mutex
+	DefById map[string]*schema.MetricDefinition
+	Tree    map[int]*Tree
```

Comment:
Avoiding pointers will not always yield better performance.  
With a basic benchmark for fetching an item from a map and modifying it, using two variants, 1 using pointers and one not, The pointer variant is 20-25% faster.
https://gist.github.com/woodsaj/40a0dec0a1b7f4109c580ef2ade0c829

```
BenchmarkPtr-4  	 5000000	       255 ns/op	     179 B/op	       0 allocs/op
BenchmarkNoPtr-4	 5000000	       323 ns/op	     179 B/op	       0 allocs/op
```

It should be noted, that maps are different to slices.  The value of a Map is not addressable.
So `data := someMap[key]`  returns a copy of value stored in the map. Any local modifications are not reflected in the value stored in the map.  To make modifications you would need to write back to the map. eg

```
data := someMap[key]
data.Updated = true
log.Printf("%t", someMap[key].Updated) // prints false
someMap[key] = data
log.Printf("%t", someMap[key].Updated) // prints true
```
However if the value is a pointer things work as expected as a copy of a pointer still gives you access to memory location the object is stored.

```
data := someMap[key]
data.Updated = true
log.Printf("%t", someMap[key].Updated) // prints true
```